Adjust task priorities to fit within supported platforms' priority ranges

pjromano opened the pull request:
| | |
|:---|:---|
|**_Related Issue(s)_**| #3985 |
|**_Has Unit Tests (y/n)_**| n |
|**_Documentation Included (y/n)_**| n |
|**_Generative AI was used in this contribution (y/n)_**| n |

---
## Change Description

Adjusted all task priorities throughout the framework so that they fit within the most-restrictive task priority range among supported platforms, which was determined to be Darwin's [15,47]. See the linked issue for that analysis. Within the fprime repository, this resulted in changes in framework-provided subtopologies and in the Ref project.

Priority changes in the fprime-tools repository are handled in [fprime-tools PR #292](https://github.com/nasa/fprime-tools/pull/292).

In-code documentation was added to clarify the intended meaning of constant `Os::Task::TASK_PRIORITY_DEFAULT`, as well as `Os::Task::TASK_DEFAULT` which follows the same pattern. The meaning has implications for derived classes' implementations which are not immediately obvious.

## Rationale

Previous task priorities were out of the valid range for the Linux and Darwin platforms, clamping tasks to the maximum priority and preventing differentiation of priority between some threads. Accordingly, priority clamp warnings were being (correctly) produced to the Logger when starting F Prime.

## Testing/Review Recommendations

No unit tests were added. The changes were verified by running the CI Integration tests and visually inspecting that the priority clamp warnings are no longer produced in the log. We should repeat this verification a final time before merging.

## Future Work

This pull request does not include documentation for recommendations on task priority. A style guide update to keep newly-added or modified threads in the framework within Darwin's range could help to prevent future code changes from reintroducing the priority clamping and provide guidance to users for their own threads.

## AI Usage (see [policy](https://github.com/nasa/fprime/blob/devel/AI_POLICY.md))

None.

pjromano committed:
Adjust task priorities to fit in supported platforms' priority ranges

Darwin's task priority range is most restrictive (15-47); adjusted priorities
to reside within that range.

pjromano committed:
Add comment clarifying TASK_PRIORITY_DEFAULT and TASK_DEFAULT

LeStarch commented on Ref/Top/RefTopology.cpp:
```diff
@@ -32,7 +32,7 @@ U32 rateGroup2Context[Svc::ActiveRateGroup::CONNECTION_COUNT_MAX] = {};
 
 enum TopologyConstants {
-    COMM_PRIORITY = 100,
+    COMM_PRIORITY = 23,
```
Comm (uplink) should be the highest priority under the rate groups, so like 34 (leaving some space between 41 and the comm).  CmdDispatch would be one step higher (35) as it consumes the output of comm, and thus should run at a higher priority.

pjromano replied:
Adjusted uplink and cmdDispatch priorities in new commit 70cca82.

LeStarch commented on Svc/Subtopologies/ComCcsds/ComCcsdsConfig/ComCcsdsConfig.fpp:
```diff
@@ -13,8 +13,8 @@ module ComCcsdsConfig {
     module Priorities {
-        constant comQueue   = 101
-        constant aggregator = 100
+        constant comQueue   = 24
```
Com queue is highest priority downlink, so it should probably be at 29, with aggregator at 30 as it "consumes" from comQueue.

pjromano replied:
Adjusted aggregator to be higher priority than comQueue, and both of these downlink tasks to be lower priority than uplink tasks, in new commit 70cca82.

LeStarch commented on Svc/Subtopologies/ComFprime/ComFprimeConfig/ComFprimeConfig.fpp:
```diff
@@ -11,7 +11,7 @@ module ComFprimeConfig {
 
     module Priorities {
-        constant comQueue   = 101
+        constant comQueue   = 24
```
29 to be consistent with the CCsds

pjromano replied:
Adjusted in new commit 70cca82.

LeStarch reviewed (commented):
This is exactly what I was hoping to see!!!

I made some comments (even though it is still in draft).  Notably in the high-priority, but w/o deadline set (uplink, downlink) I recommend adjusting the priorities such that consumers of data are higher priority than producers of data.  I left specifics.

This review was easy because you've established a clear scheme for priorities, so all I needed to do was nudge a bit. 

Thanks!

pjromano commented:
> I made some comments (even though it is still in draft). Notably in the high-priority, but w/o deadline set (uplink, downlink) I recommend adjusting the priorities such that consumers of data are higher priority than producers of data. I left specifics.
> 
> This review was easy because you've established a clear scheme for priorities, so all I needed to do was nudge a bit.

@LeStarch, thanks for the review! I'll take your updates into account in a new commit before final submission.

Note that in the changes as I originally submitted, I kept the relative order between the F Prime threads as they were before. My approach was to subtract 77 from every priority, to place the originally-used range [95,120] at [18,43], leaving some room above and below within Darwin's range [15,47]. The one exception was Ref project's blockDrv (was 140), which needed to be brought down a bit further to reach the valid range, but I still kept this as highest priority of the threads in that project, i.e. 47.

Just noting that your review makes changes to the relative priorities between the threads, while I was originally maintaining the relative priority. But your recommendation makes sense to me that consumers should be higher priority than producers to make sure consumers get to service queues so they don't sit in overflow.

pjromano committed:
Adjust relative task priorities for uplink and downlink

This moves uplink tasks to higher priority than downlink tasks and places
consumer tasks at higher priority than producer tasks.

LeStarch reviewed (approved):
Exactly what I expected.  Thanks for the extra comments as well!
